File: src/main/java/parser/Token.java
package parser;

/**
 * A token is an identifier used within a programming language. E.g. int, class, etc..
 *
 * @param <E> The tokens type enumeration.
 */
public interface Token<E extends Enum> {

    /**
     * Get the tokens type for comparison. This should be a unique number per type (ideally the ordnal of an
     * enumeration).
     *
     * @return The tokens type.
     */
    E getType();

    /**
     * Get the tokens identifier. This is the string which 'created' the token.
     *
     * @return The tokens identifier.
     */
    String getIdentifier();
}

File: src/main/scala/compiler/scanner/Token.scala
package compiler.scanner

/**
  * A token is an identifier used within a programming language. E.g. int, class, etc..
  *
  * @param identifier The identifier.
  * @param typ        The type.
  * @param line       The line.
  * @param column     The column.
  * @param source     The source.
  * @tparam E The tokens type enumeration.
  * @author Julian Schlichtholz
  * @version 1.0.0
  */
abstract class Token[E <: Enumeration#Value](val identifier: String, val typ: E, val line: Int, val column: Int, val source: String) {

  override def toString: String = {
    "%s %s at position %d:%d in %s".format(typ, identifier, line, column, source)
  }
}
